source/opt/vector_dce.cpp

```diff
@@ -0,0 +1,287 @@
+// Copyright (c) 2018 Google LLC.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "vector_dce.h"
+
+namespace {
+const uint32_t kExtractCompositeIdInIdx = 0;
+const uint32_t kInsertCompositeIdInIdx = 1;
+}  // namespace
+
+namespace spvtools {
+namespace opt {
+
+Pass::Status VectorDCE::Process(ir::IRContext* ctx) {
+  InitializeProcessing(ctx);
+
+  bool modified = false;
+  for (ir::Function& function : *get_module()) {
+    modified |= VectorDCEFunction(&function);
+  }
+  return (modified ? Status::SuccessWithChange : Status::SuccessWithoutChange);
+}
+
+bool VectorDCE::VectorDCEFunction(ir::Function* function) {
+  LiveComponentMap live_components;
+  FindLiveComponents(function, &live_components);
+  return RewriteInstructions(function, live_components);
+}
+
+void VectorDCE::FindLiveComponents(ir::Function* function,
+                                   LiveComponentMap* live_components) {
+  std::vector<WorkListItem> work_list;
+
+  // Prime the work list.  We will assume that any instruction that does
+  // not result in a vector is live.
+  //
+  // TODO: If we want to remove all of the dead code with the current
+  // implementation, we will have to iterate VectorDCE and ADCE.  That can be
+  // avoided if we treat scalar results as if they are vectors with 1 element.
```

Comment:
So, I wonder if at some point it will make better sense to give up and implement a proper memory SSA based on FUD chains so that we can do things like this and array copyprop more easily.  We've talked about this offline, and none of us are convinced.  But adding these ad-hoc passes is a long-term technical debt as well.

At some point, maybe better to use a single framework for reasoning about memory.  Passes that need it, build it, use it and tear it down.

No.  I'm not volunteering just yet ;)

Reply:
Sure, but we would need to come up with a good design for it to make sure it can be used.